### ISSUE TITLE
Allow orchestrators to call activity functions across Function app boundaries

### ISSUE THREAD

**Issue by kashimiz** [Azure/azure-functions-durable-extension#156]
**Summary**
Currently orchestrator functions and activity functions can only interact if located inside the same function app or task hub. There are situations in which it would be useful to be able to call activity functions from orchestrators in a different function app.

**Example Use Case**
A user would like to separate their activity functions into two or more Function Apps for separate management and deployment, and call them from an orchestrator in a third Function App which is managed separately. ex. aggregating results from calls to multiple APIs

**Known Challenges**
* At present there is no straightforward way to validate whether an activity function in another function app actually exists.
* Queues within a task hub are shared across all functions, so there's no way to guarantee that a message intended for FunctionAppA does not get picked up by FunctionAppB.

**Comment by SimonLuckenuik** [comment]
I agree that this would simplify some scenarios. Currently, we are forced to start another orchestration in the other function and Events to synchronize activities execution (and communicate errors). 

But on the other hand, if you consider that each Function App is equivalent to a microservice, it should be autonomous and should not rely on well known internal activity inside the other app/service, but rather use communication mechanisms with low coupling (HTTP API, Messages,...).

**Comment by cgillum** [comment]
Thank you @SimonLuckenuik, that's good feedback which we should consider.

**Comment by cgillum** [comment]
This is a common ask, so I've added the **documentation** label to this issue to track writing documentation that covers cross function app scenarios. I've also updated the original issue description with an additional challenge and a recommended workaround.

Until sing queues or HTTP is the recommended way to do cross-function app communication. You can take advantage of the async HTTP polling APIs to coordinate two orchestrations.

**Comment by cgillum** [comment]
This came up in an internal discussion, so I thought I'd post an example here using HTTP to coordinate two orchestration instances across two function apps: 

```csharp
[FunctionName("OrchestratorA")]
public static async Task RunRemoteOrchestrator(
    [OrchestrationTrigger] DurableOrchestrationContext context)
{
    // Do some work...

    // Call a remote orchestration
    string statusUrl = await context.CallActivityAsync<string>(
        "StartRemoteOrchestration", "OrchestratorB");

    // Wait for the remote orchestration to complete
    while (true)
    {
        bool isComplete = await context.CallActivityAsync<bool>("CheckIsComplete", statusUrl);
        if (isComplete)
        {
            break;
        }

        await context.CreateTimer(context.CurrentUtcDateTime.AddMinutes(1), CancellationToken.None);
    }

    // B is done. Now go do more work...
}

[FunctionName("StartRemoteOrchestration")]
public static async Task<string> StartRemoteOrchestration([ActivityTrigger] string orchestratorName)
{
    using (var response = await HttpClient.PostAsync(
        $"https://appB.azurewebsites.net/orchestrations/{orchestratorName}",
        new StringContent("")))
    {
        string statusUrl = await response.Content.ReadAsAsync<string>();
        return statusUrl;
    }
}

[FunctionName("CheckIsComplete")]
public static async Task<bool> CheckIsComplete([ActivityTrigger] string statusUrl)
{
    using (var response = await HttpClient.GetAsync(statusUrl))
    {
        // 200 = Complete, 202 = Running
        return response.StatusCode == HttpStatusCode.OK;
    }
}
```
There are many more possibilities, but hopefully this gives a good idea of how this issue could potentially be worked around.

**Comment by CarlosSardo** [comment]
@cgillum @kashimiz 

A **suggestion**, on the coordinate two orchestration instances across two function apps example: 

Integrate this example code in Durable Functions  [(DurableOrchestrationContextBase](https://github.com/Azure/azure-functions-durable-extension/blob/master/src/WebJobs.Extensions.DurableTask/DurableOrchestrationContextBase.cs)) 


```csharp
// ...
public virtual Task StartRemoteOrchestratorAsync(string functionAppUri, string orchestratorName, object input)
// ...
```

We could then use it to call/start remote orchestrators as we normally use for [Sub-Orchestrations](https://docs.microsoft.com/en-us/azure/azure-functions/durable-functions-sub-orchestrations):

```csharp
[FunctionName("ProvisionNewDevices")]
public static async Task ProvisionNewDevices(
    [OrchestrationTrigger] DurableOrchestrationContext ctx)
{
    string[] deviceIds = await ctx.CallActivityAsync<string[]>("GetNewDeviceIds");

    // Run multiple device provisioning flows in parallel, in a remote orchestrator running in a different function app 
    var provisioningTasks = new List<Task>();
    foreach (string deviceId in deviceIds)
    {
        Task provisionTask = ctx.StartRemoteOrchestratorAsync("https://appB.azurewebsites.net/orchestrations/", "RemoteOrchestrator", deviceId);
        provisioningTasks.Add(provisionTask);
    }

    await Task.WhenAll(provisioningTasks);

    // ...
}
```

**Comment by cgillum** [comment]
The challenge with this is there isn't a built-in API for starting an orchestration over HTTP. We could create such a in API, but then there's also the challenge of discovering the system function key needed to authorize that HTTP call. I don't know how we would solve that. 😕

**Comment by CarlosSardo** [comment]
How about starting it with a _POST_ to the _Http_Start_ (like in the example from the internal discussion?)

**Comment by cgillum** [comment]
In the example I posted above, I assumed that the developer had implemented their own Http_Start trigger for starting an orchestration, just like the one we ship in the samples. This is not built-in, however, which is why we can't create an API that assumes it exists.

For now, I think the best guidance is to have developers implement cross function-app themselves using whichever trigger mechanism makes sense for them (HTTP, queues, Event Grid, etc.).

I agree this could be a useful feature. If we were to build this ourselves, then we would need to satisfy the challenges that are called out at the top of this thread.

**Comment by worldspawn** [comment]
I'd like to clarify a something on durable function behaviour. If I have two functions projects, both define their own set of functions and both live in the same task hub should I be able to have functions in project A call functions in project B and vice versa?

I've just split a project out as a I have a subset of functions that are highly io constrained so I had limited the number of concurrent activity executions to 2. However there were a bunch of other functions that had no heavy io use and this limit was a burden on total throughput.

After splitting my io heavy functions into another class lib and running them both concurrently (locally) I am getting errors messages like:
``` 
Executed 'ProcessSourceFileOrchestrator' (Failed, Id=8b8ee3fd-61bc-413e-9b96-06e07d472e1f)
[29/08/2019 1:34:42 AM] System.Private.CoreLib: Exception while executing function: ProcessSourceFileOrchestrator. System.Private.CoreLib: Orchestrator function 'ProcessSourceFileOrchestrator' failed: The orchestrator function 'ProcessDrawingPdfOrchestrator' failed: "The function 'GetPdfInfo' doesn't exist, is disabled, or is not an activity function. Additional info: The following are the known activity functions: 'GetTextFromRegions', 'CreateOcrArtifact', 'AzureTextExtraction', 'EventNotify'.". See the function execution logs for additional details.
```
So durable functions just went from hero to zero in my eyes.

**Comment by cgillum** [comment]
I'm sorry to hear that Durable Functions just went from hero to zero in your eyes.

What you're describing will not work. If two functions apps share the same task hub, they will compete with each other for messages on the queue without knowing which messages are actually intended for who. If project A receives a message for a function that is defined only in project B, then project A will throw the error that you're seeing.

I think Simon put it best when he said this:

> if you consider that each Function App is equivalent to a microservice, it should be autonomous and should not rely on well known internal activity inside the other app/service, but rather use communication mechanisms with low coupling (HTTP API, Messages,...) to trigger those multi-app scenarios.

**Comment by worldspawn** [comment]
> I'm sorry to hear that Durable Functions just went from hero to zero in your eyes.

Ok slight over reaction. Sorry.

Its probably more accurate to say my issue is not with the behaviour I described but that I cannot constrain the concurrent execution of a specific activity function which causing me to look for solutions like this.

**Comment by cgillum** [comment]
@worldspawn that's fair, and I think the problem you're trying to solve is valid. Indeed, cross-function app communication would be one way to do it. Alternatively, we would need to come up with some kind of per-function concurrency limit feature.

**Comment by worldspawn** [comment]
I put this together loosely based on code on pages like this: https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-instance-management

Hope this helps someone :)

I found having this entire remote call in a sub orchestration simpler as I recreating the payload per continue become a much more simpler, focused exercise. Consuming code barely changes:

``` csharp
pdfInfo = JObject.Parse(await context.CallSubOrchestratorAsync<string>("RunRemoteOrchestration", instanceId, new {
                    orchestratorName = Constants.FunctionNames.GetPdfInfoOrchestrator,
                    payload = new
                    {
                        jobId,
                        fileId,
                        fileUri
                    }
                }));
```

Remote Execution

``` csharp
public class RemoteExecution
    {
        private readonly ILogger<RemoteExecution> _logger;
        private static readonly HttpClient Client;
        private static readonly string RemoteHost;

        static RemoteExecution(){
            Client = new HttpClient();
            RemoteHost = Environment.GetEnvironmentVariable("RemoteOrchestrationHost");
        }
        
        public RemoteExecution(ILogger<RemoteExecution> logger)
        {
            _logger = logger;
        }

        [FunctionName("RunRemoteOrchestration")]
        public async Task<string> RunRemoteOrchestration([OrchestrationTrigger] DurableOrchestrationContext context){
            var input = context.GetInput<JObject>();
            string statusQueryGetUri = null;

            if (input.SelectToken("statusQueryGetUri") == null) {
                statusQueryGetUri = await context.CallActivityAsync<string>("StartRemoteOrchestration", input);
                input.Add("statusQueryGetUri", new JValue(statusQueryGetUri));
            }
            else {
                statusQueryGetUri = input.SelectToken("statusQueryGetUri").Value<string>();
            }

            var result = await  context.CallActivityAsync<string>("CheckIsComplete", statusQueryGetUri);
            if (result == null) {
                context.ContinueAsNew(input);
                return null;
            }

            return result;
        }

        [FunctionName("StartRemoteOrchestration")]
        public async Task<string> StartRemoteOrchestration([ActivityTrigger] JObject message)
        {
            var orchestratorName = message.SelectToken("orchestratorName").Value<string>();
            var payload = message.SelectToken("payload").Value<JObject>();
            using (var response = await Client.PostAsync(
                $"http://{RemoteHost}/api/orchestrators/{orchestratorName}/wait",
                new StringContent(payload.ToString())))
            {
                var reply = await response.Content.ReadAsStringAsync();
                return JObject.Parse(reply).SelectToken("statusQueryGetUri").Value<string>();
            }
        }

        [FunctionName("CheckIsComplete")]
        public async Task<string> CheckIsComplete([ActivityTrigger] string statusUrl)
        {            
            _logger.LogDebug("Check Status Url {statusUrl}", statusUrl);
            using (var response = await Client.GetAsync(statusUrl))
            {
                if (response.StatusCode == HttpStatusCode.OK)
                {
                    var reply = await response.Content.ReadAsStringAsync();
                    var output = JObject.Parse(reply).SelectToken("output");
                    return output.ToString();
                }
                else {
                    return null;
                }
            }
        }
    }
```

On the receiving end

``` csharp
public class HttpOrchestrationTrigger
    {
        private readonly ILogger<GetPdfInfoOrchestration> _logger;

        public HttpOrchestrationTrigger(ILogger<GetPdfInfoOrchestration> logger)
        {
            _logger = logger;
        }

        [FunctionName("GetPdfInfoOrchestrator_HttpStart")]
        public async Task<HttpResponseMessage> Start(
            [HttpTrigger(AuthorizationLevel.Function, methods: "post", Route = "orchestrators/{functionName}/wait")] HttpRequestMessage req,
            [OrchestrationClient] DurableOrchestrationClient starter,
            string functionName)
        {
            var eventData = await req.Content.ReadAsStringAsync();
            var instanceId = await starter.StartNewAsync(functionName, JObject.Parse(eventData));

            _logger.LogInformation($"Started orchestration with ID = '{instanceId}'.");

            return starter.CreateCheckStatusResponse(
                req,
                instanceId);
        }
    }
```

**Comment by brandonh-msft** [comment]
Worth updating this thread w/ [this API](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-http-features#consuming-http-apis)?

**Comment by jonnyguo1988** [comment]
I would say I like @worldspawn idea , saying that Orchestration function as workflow, and activity as a microservice. Activity is stateless and it doesn't care which function executed it, put shared activities into a separate function and let it work with the _workitem_ queue in taskhub which could be possible. Orchestration functions can read the activity execute result from the workitem queue. The implement will be more simple than over http.
Calling a sub-orchestration function to another function app could not be in the field.

**Comment by ConnorMcMahon** [comment]
I think this issue is largely irrelevant as Brandon pointed out last year now that we have CallHttpAsync().

Customers can call orchestrations across app boundaries since the API knows how to navigate the HTTP async pattern.

Customers can call "activities" across app boundaries by making the activities HTTP triggered functions and calling via CallHttpAsync().

This does involve some boilerplate, but it should be minimal and can be easily avoided with helper methods.

I'm going to close the ticket in the meantime, though if people have thoughts that this work item is not complete, we can always reopen or file a new ticket to expand on this idea.
